src/porepy/numerics/ad/grid_operators.py

```diff
@@ -250,251 +268,417 @@ def __init__(
-        # The projections are wrapped by a pp.ad.SparseArray to be compatible with the
-        # requirements for processing of Ad operators.
-        def bmat(matrices, name):
-            # Create block matrix, convert it to optimized storage format
-            block_matrix = pp.matrix_operations.optimized_compressed_storage(
-                sps.bmat(matrices)
-            )
-            return SparseArray(block_matrix, name=name)
+            # Here we need to store both intensive and extensive projections.
+            self._mortar_to_primary_int: Optional[SparseArray] = None
+            self._mortar_to_primary_avg: Optional[SparseArray] = None
+            self._primary_to_mortar_int: Optional[SparseArray] = None
+            self._primary_to_mortar_avg: Optional[SparseArray] = None
+
+        # The case for secondary grids is analogous to the primary grids.
+        if self._is_conforming_secondary:
+            self._mortar_to_secondary: Optional[SparseArray] = None
+            self._secondary_to_mortar: Optional[SparseArray] = None
+        else:
+            self._mortar_to_secondary_int: Optional[SparseArray] = None
+            self._mortar_to_secondary_avg: Optional[SparseArray] = None
+            self._secondary_to_mortar_int: Optional[SparseArray] = None
+            self._secondary_to_mortar_avg: Optional[SparseArray] = None
+
+    def sign_of_mortar_sides(self) -> SparseArray:
+        """Construct a matrix that gives the sign of the mortar sides.
+
+        The matrix is a block diagonal matrix, where each block corresponds to a mortar
+        grid.
+
+        Returns:
+            Sparse matrix (in the Ad sense) that represents the sign of the mortar
+            sides.
+
+        """
+        # Shortcut for the case where there are no interfaces.
+        if len(self._interfaces) == 0:
+            return SparseArray(sps.bmat([[None]]), name="SignOfMortarSides")
+
+        mats = []
+        for intf in self._interfaces:
+            assert isinstance(intf, pp.MortarGrid)  # Appease mypy
+            mats.append(intf.sign_of_mortar_sides(self.dim))
+        else:
+            return SparseArray(sps.block_diag(mats), name="SignOfMortarSides")
+
+    def mortar_to_primary_int(self) -> SparseArray:
+        """Construct a matrix that projects from mortar grids to primary grids for
+        extensive quantities.
+
+        Returns:
+            Sparse matrix (in the Ad sense) that represents the projection.
+
+        """
 
-        self.mortar_to_primary_int = bmat(
-            [mortar_to_primary_int], name="MortarToPrimaryInt"
+        # Retrieved cached projection if it exists. EK note: I tried to include this
+        # logic also in the helper method _construct_projection, but that led to too
+        # complex logic, so I found it better to keep it here.
+        if self._is_conforming_primary and self._mortar_to_primary is not None:
+            return self._mortar_to_primary
```

Comment:
Do I see correctly that there is the possibility (in the case of conforming grids), that one receives an operator which carries the name "MortarToPrimaryAvg" here, if that one was requested first. If this is not desired, one could consider assigning names withough "Avg" and "Int" in case of conforming grids.

Reply:
Fair point, see response to @IvarStefansson's similar remark just below.